File: Practica_2/Ejercicio_3/Algebra_Vectorial.py
import math
from multimethod import multimethod
class Algebra_vectorial:
    def __init__(self, x=0, y=0, z=0):
        self.__x = x
        self.__y = y
        self.__z = z

    def getX(self):
        return self.__x

    def getY(self):
        return self.__y

    def getZ(self):
        return self.__z

    # a) Suma: c = a + b
    def __add__(self, b: 'Algebra_vectorial'):
        return Algebra_vectorial(self.__x + b.getX(), self.__y + b.getY(), self.__z + b.getZ())

    #b) Multiplicación por escalar: b = ra
    @multimethod
    def __mul__(self, r = float):
        return Algebra_vectorial(r * self.__x, r * self.__y, r * self.__z)

    # c) Longitud magnitud: |a|
    def magnitud(self):
        return math.sqrt(self.__x ** 2 + self.__y ** 2 + self.__z ** 2)

    # d) Normal: b = a/|a|
    def __truediv__(self, b: 'Algebra_vectorial'):
        m = self.magnitud()
        if (m == 0): return Algebra_vectorial()
        return Algebra_vectorial(self.__x / m, self.__y / m, self.__z / m)

    # e) Producto escalar: a * b
    @multimethod
    def __mul__(self, b: 'Algebra_vectorial'):
        return (self.__x * b.getX()) + (self.__y * b.getY()) + (self.__z * b.getZ())

    # f) Producto vectorial cruz: a x b
    def  __pow__(self, b: 'Algebra_vectorial'):
        cx = self.__y * b.getZ() - self.__z * b.getY()
        cy = self.__z * b.getX() - self.__x * b.getZ()
        cz = self.__x * b.getY() - self.__y * b.getX()
        return Algebra_vectorial(cx, cy, cz)

    def __str__(self):
        return f'({self.__x}, {self.__y}, {self.__z})'

a = Algebra_vectorial(1,0,0)
b = Algebra_vectorial(0,1,0)
# a) Suma: c = a + b
print("Suma (a + b): ", a + b);
# b) Multiplicación por escalar: b = ra
print("Escalar (a * 5): ", a * 5.0)
# c) Longitud magnitud: |a|
print("Longitud (|a|): ", a.magnitud())
# d) Normal: b = a/|a|
print("Normal (a/|a|): ", a / a)
# e) Producto escalar: a * b
print("Punto (a * b): ", a * b)
# f) Producto vectorial cruz: a x b
print("Cruz (a x b): ", a ** b)

File: Practica_2/Ejercicio_3/AlgebraVectorial.java
package Ejercicio_3;

public class AlgebraVectorial {
    private double x, y, z;

    public AlgebraVectorial() {
        this.x = 0; 
        this.y = 0; 
        this.z = 0;
    }

    public AlgebraVectorial(double x, double y, double z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public double getX() {
    	return x; 
    }
    public double getY() {
    	return y; 
    }
    public double getZ() {
    	return z; 
    }

    // a) Suma: c = a + b
    public AlgebraVectorial sumar(AlgebraVectorial b) {
        return new AlgebraVectorial(this.x + b.getX(), this.y + b.getY(), this.z + b.getZ());
    }

    // b) Multiplicación por escalar: b = ra
    public AlgebraVectorial multiplicar(double r) {
        return new AlgebraVectorial(r * this.x, r * this.y, r * this.z);
    }

    // c) Longitud magnitud: |a|
    public double magnitud() {
        return Math.sqrt(x * x + y * y + z * z);
    }

    // d) Normal: b = a/|a|
    public AlgebraVectorial normalizar() {
        double m = this.magnitud();
        return new AlgebraVectorial(this.x / m, this.y / m, this.z / m);
    }
    
    // e) Producto escalar: a * b
    public double multiplicar(AlgebraVectorial b) {
        return (this.x * b.getX()) + (this.y * b.getY()) + (this.z * b.getZ());
    }

    // f) Producto vectorial cruz: a x b
    public AlgebraVectorial productoCruz(AlgebraVectorial b) {
        double cx = this.y * b.getZ() - this.z * b.getY();
        double cy = this.z * b.getX() - this.x * b.getZ();
        double cz = this.x * b.getY() - this.y * b.getX();
        return new AlgebraVectorial(cx, cy, cz);
    }

    @Override
    public String toString() {
        return "(" + x + ", " + y + ", " + z + ")";
    }
}